torchgen/gen.py

```diff
@@ -1374,6 +1345,81 @@ def flatten_pre_group(
     )
 
 
+# Return native function declarations grouped by their namespaces.
+def get_native_function_declarations(
+    *,
+    grouped_native_functions: Sequence[Union[NativeFunction, NativeFunctionsGroup]],
+    backend_indices: Dict[DispatchKey, BackendIndex],
+) -> List[str]:
+    declarations: List[str] = []
+    ns_grouped_kernels: Dict[str, List[str]] = defaultdict(list)
+    newline = "\n"
+    for f in grouped_native_functions:
+        for backend_idx in backend_indices.values():
+            backend_metadata = backend_idx.get_kernel(f)
+            namespace = (
+                backend_metadata.cpp_namespace
+                if backend_metadata
+                else DEFAULT_KERNEL_NAMESPACE
+            )
+            ns_grouped_kernels[namespace].extend(
+                dest.compute_native_function_declaration(f, backend_idx)
+            )
+
+    for namespace, kernels in ns_grouped_kernels.items():
```

Comment:
Actually, I think this might create duplicate declarations (see https://github.com/pytorch/pytorch/pull/79733/files#r915842429). `kernels` is just a list of entries (one generated for every valid `(NativeFunction, BackendIndex)` pair), so it might have more than one entry for a given `NativeFunction`.

The comment basically suggests fixing that by taking your list (`kernels`), converting to a set to remove duplicates (using `OrderedDict.from_keys` to the codegen output is deterministic), and converting back to a list